Unref uv timers/tcp handles

kpayson64 opened the pull request:
Fixes #14947.
 
@mehrdada This is a release blocker

During #14599 there was some offline discussion regarding the unrefing of uv_handles in gRPC.
At the time, we weren't sure exactly why we were doing it, and when I removed the unrefing it appeared that all the node tests and c-core libuv tests were passing, so I left it out.

It turns out this is very much necessary.  Node uses a garbage collector, so these timers/channels will block the uv loop from exiting until they are garbage collected.

The failures were not noticed immediately because they don't show up in the node unit tests, only the interop tests.  The Node interop tests that get run on PRs don't use the pending changes to c-core, so I did not notice the failures at the time.

mehrdada commented:
Do we need to ship an RC3 for all languages or just Node?

kpayson64 commented:
@mehrdada 
Just Node

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

mehrdada commented:
If Node hasn't published its RC2 yet, I think we should just build at head of 1.11.x and publish it as RC2 for Node, increment the version to final on Friday, ship final packages for every other language right then, and then Node can wait a bit to see things settled down and publish it as final. Sounds reasonable?

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

nicolasnoble commented:
Can you please explain to me what happens if you don't have the `#ifndef` ? More specifically, why wouldn't we be able to always unref timers after their creation, like what the original code was doing ?

nicolasnoble commented:
For reference: https://github.com/grpc/grpc/blob/v1.10.x/src/core/lib/iomgr/timer_uv.cc#L71

kpayson64 commented:
After the refactor, I simplified the custom pollset implementation to only have a global shutdown method.  

Before, there was also an individual pollset shutdown function that was only used when running the c-core libuv tests that would spin the loop to make sure all callbacks were run.

After the refactor, I switched the global shutdown to spin the libuv loop until there were no active/referenced uv objects.
 
```#ifdef GRPC_UV_HOLD_LOOP```
is only defined for the c-core tests.

nicolasnoble commented:
I understand that part, but what I don't see is why we couldn't unconditionally unref the timers even during tests. What breaks exactly if we do ? Timer-only tests ?

kpayson64 commented:
When I unconditionally unrefed the tcp endpoints, the uv-loop exits before all tcp-shutdown callbacks are called (which caused memory leak asan failures).

I suspect something similar would happen for timers.

nicolasnoble commented:
So, what I am getting at is that by doing something one way in node, and another way in tests, we're putting ourselves in a situation where we're not testing what actually happens in reality. What I am inferring from your last response is that if doing this unref can cause memory leaks in tests, doing it in the node codepath only might cause memory leaks in node that will then go undetected because we're not doing it when testing for memory leaks.

Moreover, I am also getting the feeling that this is caused by the global pollset shutdown change, indicating that the rootcause of this is in fact in there, and that it wasn't the right thing to do to begin with. This is where the fix needs to happen, not with this workaround.

Or to be more blunt, you can't keep adding kludges to fix other kludges.

mehrdada commented:
@nicolasnoble Shall we skip a Node release in 1.11?  I am concerned about a rushed release after deeper fixes in core without going to a proper QA process. The next branch cut (1.12) is April 23 and 1.12 is planned to be released on May 7.  I am partial to skipping Node 1.11 and going with a release from the head of 1.11 as is without this PR. What do you think?

nicolasnoble commented:
So my short answer here is "no". There are lots of changes in grpc/grpc-node that our users are waiting for, and some of our future release schedule depend on features being released now.

Now, I am going to suggest something potentially controversial, but since grpc/grpc-node is its own separate repository, grpc-node could still be released for version 1.11, but with the grpc-core submodule being pointed at 1.10. This might create the possibility that some of the grpc-node users will be puzzled as to why some grpc-core bugfixes in 1.11 aren't present in grpc-node 1.11, but this would allow us to move forward nonetheless.

mehrdada commented:
@nicolasnoble That sounds an acceptable resolution to me, especially since 1.12 feature window will be quite short anyway.
